### PR TITLE
[8.17] Backport of [coqide] Revert behavior changes in offset location

### PR DESCRIPTION
**Pull request by ejgallego**
Backport of #17348

### CODE REVIEW

**Comment by jfehrle** [comment]
I tested the following cases:
- debugger breakpoints work as expected
- Pff.v takes about 3 minutes to run.  Tooltips for the warnings and errors appear correctly
- @ppedrot's [test](https://github.com/coq/coq/pull/17382#issuecomment-1470002594) (tooltips generated after modifying a failed proof in async mode are correctly positioned).
- tooltips are correct when Unicode characters are present

Are there any other cases I've missed?

File: ide/coqide/idetop.ml
Comment (ejgallego) [reply ide/coqide/idetop.ml:95]: See my comment at https://github.com/coq/coq/pull/17348#discussion_r1139215802

**Comment by ejgallego** [comment]
This reopens https://github.com/coq/coq/issues/16987 + unreported "bad error number in panel"

**Comment by jfehrle** [comment]
That's the case in which loc is not passed in the XML.  IIRC that was quite a simple fix, should be easy and quick to include in this PR.

**Comment by Zimmi48** [comment]
IIUC this is OK to merge?

**Comment by ejgallego** [comment]
> An information which I'm also missing is whether choosing one or the other PR for 8.17 has an impact on coq-lsp?

It has no impact.

I think @herbelin your summary is correct.

Actually I didn't want to bring this topic on a 8.17 release setting, but indeed after any of the PRs we are talking purely about a CoqIDE issue. So it would be possible to release CoqIDE 8.17.1 etc... independently with the changes to bring the fixes from #17382 , note how that PR touches nothing in coq-core / coq-stdlib.

#17348 only adds a "hook" to allow us to continue the old hack, but indeed I think there is limited value to keeping the hack, but indeed we got ourselves in a strange situation here.

**Comment by jfehrle** [comment]
> IIUC this is OK to merge?

I'd like to take a closer look at the code.  Some of the details don't make sense at the moment.

File: ide/coqide/idetop.ml
Comment (jfehrle) [reply ide/coqide/idetop.ml:95]: @ejgallego See my reply at the same location.

**Comment by jfehrle** [comment]
> Sorry, I'm lost. Isn't it either https://github.com/coq/coq/pull/17381 or https://github.com/coq/coq/pull/17382?

@ejgallego Opening multiple very similar PRs in multiple branches is totally confusing.  It would be been a lot clearer to put out a single PR for 8.17 and get that reviewed/merged before starting a discussion of a PR for 8.18.

See my comment https://github.com/coq/coq/pull/17348#discussion_r1139535447.  Creating `fix_loc` is probably a mistake, but I won't object if @Zimmi48 chooses to merge this as-is.

**Comment by jfehrle** [comment]
See https://github.com/coq/coq/pull/17348#discussion_r1139686581

**Comment by Zimmi48** [comment]
>Sorry, I'm lost. Isn't it either #17381 or #17382?

I thought that #17382 was only considered for master because of the change in the XML protocol. But if you think that it has been tested enough and it is not a breaking change, then yes, we could consider it for 8.17.0 as well, instead of this PR. In any case, I won't take the decision instead of the CoqIDE maintainers.

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : how about you merge the things you see fit into 8.17, you notify us and then we test the 8.17 branch. I find testing multiple branches highly confusing and not very effective. If you merge first and we test then, we test what gets released. In the unlikely case that there are issues we can revert. For a GUI app I find this approach more reasonable than the typical "first CI then merge" approach - which makes more sense if tests can be automated. For a GUI App IMHO one should optimise the process for test effectiveness.

**Comment by Zimmi48** [comment]
If there is a choice between two solutions, I don't see how this would work. But if you want me to take an arbitrary decision without knowing that part of the code well, then I would guess this PR is the safest option for 8.17 and thus merge this one.

**Comment by MSoegtropIMC** [comment]
You can easily do the math of what likelihood you have to find issues if few people test individual branches upfront or if more people test a merged "release candidate". If you take into account the possibilities that both branches are OK and the likelihood of finding issues in "random user tests", for a wide range of assumed probabilities the answer should be that testing after merging with more people has a better probability of finding issues.

I didn't do the math here, but I did such computations frequently in the past and my gut feeling for the outcome is quite clear.

**Comment by MSoegtropIMC** [comment]
So what I would do is a as well informed as possible decision based on the available information - possibly explicitly asking specific people for more information - merge and request the test.

**Comment by jfehrle** [comment]
> @Zimmi48 But if you want me to take an arbitrary decision without knowing that part of the code well, then I would guess this PR is the safest option for 8.17 and thus merge this one.

Agreed.  AFAIU #17382 doesn't fix any additional issues and or add anything urgent, so it doesn't need to be done now.  Several people have already tested this PR across multiple environments.

> @MSoegtropIMC how about you merge the things you see fit into 8.17, you notify us and then we test the 8.17 branch.

Sounds good to me.

**Review by jfehrle — APPROVED** [review]

**Comment by herbelin** [comment]
My conclusion/understanding being that both solutions are ok (or, at worst, that #17382 also solved the not-8.17-new location shift problem when inserting characters while processing).

In particular, I'm ok with @jfehrle's choice.

**Comment by Zimmi48** [comment]
OK, then I merge this now and start preparing the release of 8.17.0. Additional testing of `v8.17` after this is merged is of course welcome, but I will try avoiding delaying 8.17.0 too much now. The inclusion in the upcoming Coq Platform beta pick will then give us more user testing.

**Comment by jfehrle** [comment]
Should we merge the same code into master to keep the branches in sync?  If that's still possible.

**Comment by ejgallego** [comment]
> Agreed. AFAIU #17382 doesn't fix any additional issues and or add anything urgent, so it doesn't need to be done now.

It does indeed fix at least two reported issues #17023 #16987 , so it is preferred for master.

It also fixes some other issues with the code that I found when doing the first original PR.

**Pull request merged**
